### DISCUSSION TITLE
Best way to iterate through the fields of the model for setting the fields of models that are downstream

### DISCUSSION THREAD

**Discussion by caniko** [pydantic/pydantic#7333] [Question] [answered]
Consider the following:
```python
from pydantic import BaseModel
from pydantic.dataclasses import dataclass


@dataclass
class A:
    _s: bool = False


class B(BaseModel):
    m: A
    y: A


my_b = B(m=A(), y=A())
```
I want to iterate the fields of B to set `_s` of `m` and `y` to `True`. In heterogeneus cases I would check the field type, and skip if it is not the right type. Doing the following, won't work because all downstream models get dumped:
```python
for f in my_b.model_dump().values():
    f._s = False
```
Raises error, because `f` is a dict from the dumping itself, and no longer the original object instance.

I am using this method for now:
```python
for f in my_b.model_fields_set:
    my_b.__getattribute__(f)._s = True
```

Any other ideas are very welcome!

**Comment by sydney-runkle** [comment, marked as answer]
The following solution is a bit cleaner.

```py
for f in my_b.model_fields_set:
    getattr(my_b, f)._s = True

print(my_b.model_dump())
# {'m': {'_s': True}, 'y': {'_s': True}}
```

Or even cleaner:
```py
for _, val in iter(my_b):
    val._s = True

print(my_b.model_dump())
# {'m': {'_s': True}, 'y': {'_s': True}}
```

**Reply by IvanivAnton** [reply]
The cleanest way:

```python
for _, val in my_b:
    val._s = True
```
Methods `__iter__`, `__next__` called under hood

Tell me if I wrong

**Comment by alexpovel** [comment]
I had a very similar issue around `model_dump`. My use case is [logging](https://www.structlog.org/en/stable/). Through some setup, I had set `repr=False` on a bunch of fields I didn't care for. That works for string representations, but I also wanted structured logging, aka dumping the models to JSON. But that ignores `repr=False`, so those fields are found in the output, congesting the logs. `exclude=True` is not an option, as I needed those fields eventually. So structured logging strikes some awkward middle ground between string representation and model dumping!

My current solution is:

```python
import logging
import typing as t

import structlog
from pydantic import BaseModel


def _dump_models(
    logger: structlog.typing.WrappedLogger,
    log_method: str,
    event_dict: structlog.typing.EventDict,
) -> structlog.typing.EventDict:
    """Dumps any found `pydantic.BaseModel`s to JSON, so logs are properly structured.

    Turns log entries from:

    ```json
    {"event": "SomeModel(a=1, b=2"}
    ```

    to:

    ```json
    {"event": {"a": 1, "b": 2}}
    ```

    That is, normal `pydantic.BaseModel` representation is a string and incompatible
    with further JSON processing.
    """
    return {k: _dump(v) for k, v in event_dict.items()}


def _dump(item: t.Any) -> t.Any:
    match item:
        case BaseModel() as model:
            subitems: dict[str, t.Any] = dict()
            fields = model.model_fields

            for field, value in model:
                if isinstance(value, BaseModel):
                    subitems[field] = _dump(value)
                    continue

                if field in fields and fields[field].repr:
                    subitems[field] = _dump(value)

            return subitems
        case list() | tuple() | set() as container:  # pyright: ignore
            # Pyright finds this disgusting; this passes `mypy` though. `  # type:
            # ignore` would fail `mypy` is it'd be unused (because there's nothing to
            # ignore because `mypy` is content)
            return type(container)(  # pyright: ignore
                _dump(i) for i in container  # pyright: ignore
            )
        case dict():
            return {
                k: _dump(v)
                for k, v in item.items()  # pyright: ignore[reportUnknownVariableType]
            }
        case _:
            return item
```

where `_dump_models` is a [processor](https://www.structlog.org/en/stable/processors.html). That's specific to structlog, but `_dump` in itself might be helpful in the general case. It recursively dumps to JSON/`dict`, respecting `repr=False`.

Not tested very well! And those ugly type-ignore comment probably have a better workaround (this passes mypy though... it's mainly `Any` anyway).
